src/mono/mono/metadata/object.c

```diff
@@ -1588,7 +1588,7 @@ build_imt_slots (MonoClass *klass, MonoVTable *vt, gpointer* imt, int slot_num)
 						vt_slot ++;
 					continue;
 				}
-				if (mono_method_get_imt_slot (method) != slot_num) {
```

Comment:
it's frustrating that `build_mit_slots` essentially duplicates the work that we do trying to assign vtable slots to methods in vtable setup.  At the very least it feels like we should be able to assert that they agree.   I tried to add something in https://github.com/dotnet/runtime/commit/e2f058cf362703085e3f425a8e4a9fb07a9242e3 but I couldn't finish it. I wonder if the issue was precisely these private methods.

Also, i wonder if there's something we can assert into common_call_trampoline when we do an imt lookup - (for example when we get back a vtable slot from imt lookup, maybe we can range check it to make sure that it is in bounds for the current klass)

Reply:
Agree, there is a lot of scenarios covered in here as well, for this scenario, the value that we read from vtable was not really used in the end, since we compiled the method based on other lookups, so it was mainly an assert that made sure that the vtable slot was not NULL, any other value would just "work", unless we would end up reading protected memory.